app/main/api/handlers/importSeed.ts

```diff
@@ -0,0 +1,218 @@
+import { AppStateManager } from "app/main/appState"
+import * as mnemonic from "app/main/crypto/mnemonic"
+
+import { AppStateS } from "app/main/system"
+import { fanOut, inject } from "app/main/utils/utils"
+import * as crypto from "crypto"
+import { LiteralType } from "io-ts"
+import { Seed } from "app/common/runtimeTypes/storage/wallets"
+import * as Slip32 from "slip32"
+import { fromMnemonics } from "app/main/crypto/seed"
+
+/**
+ * Handler function for `importSeed` method.
+ * It receives a param with the mnemonic words and check they are valid.
+ * If there is an `UnconsolidatedWallet` in the app state it will also validate that they are the
+ * same.
+ * If there is no `UnconsolidatedWallet` in the app state, it will create one with the mnemonics.
+ * In both cases it will also generate a wallet ID and return it.
+ */
+export default async function importSeed
+  ({ appStateManager }: AppStateS, params: any): Promise<JsonSerializable> {
+
+  // First of all, parse method parameters
+  return parseParams(params)
+    // Get the seed from the params
+    .then(inject(getSeed, appStateManager))
+    .then(buildUnconsolidatedWallet)
+    // .then(fanOut([validateMnemonicsString, inject(matchMnemonics, appStateManager), generateId]))
+    // Update or insert unconsolidated wallet into app state
+    .then(inject(upsertUnconsolidated, appStateManager))
+    // The handler logic ends here. What follows is response building and encoding.
+    // If everything was OK, return success
+    .then(buildSuccessResponse)
+    // If anything failed, return error
+    .catch(buildErrorResponse)
+    // Encode response
+    .then(encodeResponse)
+}
+
+// Function to assert a never value
+const assertNever = (x: never) => undefined
+
+/**
+ * Get seed from params
+ * @param params
+ * @param appStateManager
+ */
+async function getSeed(params: ImportSeedParams, appStateManager: AppStateManager): Promise<Seed> {
+  let seed: Promise<Seed>
+  switch (params.kind) {
+    case "mnemonics":
+      seed = processMnemonics(params.mnemonics, appStateManager)
+      break
+    case "xprv":
+      seed = processXprv(params.xprv)
+      break
+    default:
+      // compiler instruction to ensure exhaustiveness
+      assertNever(params)
+      throw importSeedErrors.INVALID_METHOD_PARAMS
+  }
+
+  return seed
+}
+
+/**
+ * Validate mnemonics and get seed
+ * @param mnemonics
+ * @param appStateManager
+ */
+async function processMnemonics(mnemonics: string, appStateManager: AppStateManager):
+  Promise<Seed> {
+  return Promise.resolve(mnemonics)
+    // In parallel: validate mnemonic, check mnemonic against unconsolidated wallet and generate id
+    .then(fanOut([validateMnemonicsString, inject(matchMnemonics, appStateManager)]))
+    .then(seedFromMnemonics)
+}
+
+/**
+ * Generates an ID and builds and returns an UnconsolidatedWallet
+ * @param seed
+ */
+async function buildUnconsolidatedWallet(seed: Seed) {
+  return {
+    id: generateId(seed),
```

Comment:
We should not generate the Id here, that's `encryptWallet's` business now! :wink: